DependencyInjection/CmfResourceExtension.php

```diff
@@ -14,143 +14,92 @@
+
+            if (!isset($this->repositoryFactories[$type])) {
+                throw new InvalidArgumentException(sprintf(
+                    'Unknown repository type "%s", known repository types: "%s"',
+                    $type,
+                    implode('", "', $repositoryTypes)
                 ));
             }
 
-            $definition = $this->{$methods[$repository['type']]}($repository['options'], $alias);
-            $definition->addTag('cmf_resource.repository', array('type' => $repository['type'], 'alias' => $alias));
-            $container->setDefinition('cmf_resource.repository.' . $alias, $definition);
-        }
-
-        $container->setAlias('cmf_resource.registry', 'cmf_resource.registry.container');
-    }
-
-    private function createDoctrinePhpcrOdmRepository(array $options, $alias)
-    {
-        if (!isset($options['basepath'])) {
-            $options['basepath'] = null;
-        }
-
-        $definition = new Definition('Symfony\Cmf\Component\Resource\Repository\PhpcrOdmRepository');
-        $definition->addArgument(new Reference('doctrine_phpcr'));
-        $definition->addArgument($options['basepath']);
-
-        unset($options['basepath']);
-
-        $this->validateRemainingOptions($options, array('basepath'), $alias);
-
-        return $definition;
-    }
-
-    private function createDoctrinePhpcrRepository(array $options, $alias)
-    {
-        if (!isset($options['basepath'])) {
-            $options['basepath'] = null;
-        }
+            $factory = $this->repositoryFactories[$type];
 
-        $definition = new Definition('Symfony\Cmf\Component\Resource\Repository\PhpcrRepository');
-        $definition->addArgument(new Reference('doctrine_phpcr.session'));
-        $definition->addArgument($options['basepath']);
+            $optionsResolver = new OptionsResolver();
+            $factory->configure($optionsResolver);
 
-        unset($options['basepath']);
-
-        $this->validateRemainingOptions($options, array('basepath'), $alias);
-
-        return $definition;
-    }
-
-    private function createCompositeRepository(array $options, $alias)
-    {
-        if (!isset($options['mounts'])) {
-            throw new InvalidConfigurationException('The composite repository type requires a "mounts" option to be set.');
-        }
-
-        $definition = new Definition('Puli\Repository\CompositeRepository');
-
-        foreach ($options['mounts'] as $mount) {
-            if (!isset($mount['mountpoint']) || !isset($mount['repository'])) {
-                throw new InvalidConfigurationException('The "mounts" option of the composite repository type requires a "mountpoint" and "repository" options to be set.');
+            try {
+                $config = $optionsResolver->resolve($config['options']);
+            } catch (\Exception $e) {
+                throw new InvalidArgumentException(sprintf(
+                    'Invalid configuration for repository "%s"',
```

Comment:
Maybe slightly controversial - I am wrapping any exceptions thrown by the OptionsResolver in order to indicate on which repository the error occurred.